src/Common/FileSegment.cpp

```diff
@@ -492,8 +502,12 @@ void FileSegment::complete(std::lock_guard<std::mutex> & cache_lock)
     if (download_state == State::SKIP_CACHE || detached)
         return;
 
-    if (download_state != State::DOWNLOADED && getDownloadedSize(segment_lock) == range().size())
+    if (isDownloaderImpl(segment_lock)
```

Comment:
The fix is in this line.
There could be a race: in this place anyone could mark the segment as downloaded in case downloaded_size equals file segment size. This is possible in case downloader updated downloaded size but did not yet call `completeFileSegmentAndResetDownloader`.

Reply:
Maybe add it as comment?